pagecache/pagecache.go

```diff
@@ -0,0 +1,360 @@
+// Package pagecache provides the implementation of a caching layer for files
+// implementing the io.ReaderAt interface.
+package pagecache
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"hash/maphash"
+	"io"
+	"math/bits"
+	"sync"
+
+	"github.com/segmentio/datastructures/v2/cache"
+)
+
+const (
+	// DefaultPageSize is the default page size used when creating a Cache
+	// instance.
+	DefaultPageSize = 4096
+
+	// DefaultPageCount is the default page count used when creating a Cache
+	// instance.
+	DefaultPageCount = 16384
+)
+
+const (
+	// The number of buckets in a Cache instance.
+	//
+	// At this time, this is a constant value, tho it may be interesting to
+	// make it configurable in the future. Having this value be a constant and
+	// a power of two allows the compiler to optimize modulo operations using
+	// bit masks, which are instructions that tend to be orders of magnitude
+	// faster. If we make this value configurable, we might want to ensure that
+	// we retain the same performance characteristics, which would require to
+	// only allow powers of two as bucket counts, and implement the bitwise
+	// optimizations in the code.
+	numBuckets = 64
+)
+
+var (
+	// ErrNoPageps is returned when memory prossure is too high and it is not
+	// possible to read files through the page cache.
+	ErrNoPages = errors.New("there are no free pages left in the cache")
+)
+
+// Config carries the configuration for the page cache.
+type Config struct {
+	PageSize  int64
+	PageCount int64
+}
+
+// DefaultConfig constructs a new Config instance initialized with the default
+// configuration.
+func DefaultConfig() *Config {
+	return &Config{
+		PageSize:  DefaultPageSize,
+		PageCount: DefaultPageCount,
+	}
+}
+
+// Apply applies the list of options passed as arguments to c.
+func (c *Config) Apply(options ...Option) {
+	for _, opt := range options {
+		opt.Configure(c)
+	}
+}
+
+// Option is an interface implemented by options allowing configuration of new
+// Cache instances.
+type Option interface {
+	Configure(*Config)
+}
+
+type option func(*Config)
+
+func (opt option) Configure(config *Config) { opt(config) }
+
+// PageSize is a cache configuration option setting the size of individual pages
+// in a Cache instance.
+//
+// If it is not a power of two, the size will be adjusted to the nearest one.
+//
+// Default: 4 KiB
+func PageSize(size int64) Option {
+	return option(func(config *Config) { config.PageSize = size })
```

Comment:
Feels like if we need a power of 2, we should fix the value here as soon as possible rather than passing around a possibly incorrect value to other parts of the code base.

Reply:
It seems we would need to do a second pass anyways for `NewWithConfig` then?